File: apps/web/src/lib/api/backups.ts
import { apiFetch } from "./core"

export interface ApiBackupConfig {
  id: string
  service_id: string
  storage_integration_id: string
  schedule: string
  retention_days: number
  path_prefix: string
  enabled: boolean
  last_backup_at: string | null
  last_backup_status: "pending" | "running" | "success" | "failed" | null
  created_at: string
  updated_at: string
}

export interface ApiSystemBackupConfig {
  id: string
  organization_id: string
  storage_integration_id: string
  schedule: string
  retention_days: number
  path_prefix: string
  enabled: boolean
  last_backup_at: string | null
  last_backup_status: "pending" | "running" | "success" | "failed" | null
  created_at: string
  updated_at: string
}

export interface BackupObject {
  key: string
  size: number
  last_modified: string
}

export interface CreateBackupBody {
  storage_integration_id: string
  schedule: string
  retention_days?: number
  path_prefix?: string
}

export interface UpdateBackupBody {
  schedule?: string
  retention_days?: number
  path_prefix?: string
  enabled?: boolean
}

export const backups = {
  list: (orgId: string, projectId: string, serviceId: string, token: string) =>
    apiFetch<ApiBackupConfig[]>(
      `/api/v1/orgs/${orgId}/projects/${projectId}/services/${serviceId}/backups`,
      {},
      token
    ),

  create: (orgId: string, projectId: string, serviceId: string, body: CreateBackupBody, token: string) =>
    apiFetch<ApiBackupConfig>(
      `/api/v1/orgs/${orgId}/projects/${projectId}/services/${serviceId}/backups`,
      { method: "POST", body: JSON.stringify(body) },
      token
    ),

  update: (orgId: string, projectId: string, serviceId: string, id: string, body: UpdateBackupBody, token: string) =>
    apiFetch<ApiBackupConfig>(
      `/api/v1/orgs/${orgId}/projects/${projectId}/services/${serviceId}/backups/${id}`,
      { method: "PATCH", body: JSON.stringify(body) },
      token
    ),

  delete: (orgId: string, projectId: string, serviceId: string, id: string, token: string) =>
    apiFetch<void>(
      `/api/v1/orgs/${orgId}/projects/${projectId}/services/${serviceId}/backups/${id}`,
      { method: "DELETE" },
      token
    ),

  trigger: (orgId: string, projectId: string, serviceId: string, id: string, token: string) =>
    apiFetch<ApiBackupConfig>(
      `/api/v1/orgs/${orgId}/projects/${projectId}/services/${serviceId}/backups/${id}/trigger`,
      { method: "POST" },
      token
    ),

  getSystem: (orgId: string, token: string) =>
    apiFetch<ApiSystemBackupConfig | null>(
      `/api/v1/orgs/${orgId}/system-backup`,
      {},
      token
    ),

  upsertSystem: (orgId: string, body: { storage_integration_id: string; schedule: string; retention_days?: number; path_prefix?: string; enabled: boolean }, token: string) =>
    apiFetch<ApiSystemBackupConfig>(
      `/api/v1/orgs/${orgId}/system-backup`,
      { method: "PUT", body: JSON.stringify(body) },
      token
    ),

  deleteSystem: (orgId: string, token: string) =>
    apiFetch<void>(
      `/api/v1/orgs/${orgId}/system-backup`,
      { method: "DELETE" },
      token
    ),

  triggerSystem: (orgId: string, token: string) =>
    apiFetch<ApiSystemBackupConfig>(
      `/api/v1/orgs/${orgId}/system-backup/trigger`,
      { method: "POST" },
      token
    ),

  listObjects: (orgId: string, projectId: string, serviceId: string, id: string, token: string) =>
    apiFetch<BackupObject[]>(
      `/api/v1/orgs/${orgId}/projects/${projectId}/services/${serviceId}/backups/${id}/objects`,
      {},
      token
    ),

  restore: (orgId: string, projectId: string, serviceId: string, id: string, key: string, token: string) =>
    apiFetch<void>(
      `/api/v1/orgs/${orgId}/projects/${projectId}/services/${serviceId}/backups/${id}/restore`,
      { method: "POST", body: JSON.stringify({ key }) },
      token
    ),

  listSystemObjects: (orgId: string, token: string) =>
    apiFetch<BackupObject[]>(
      `/api/v1/orgs/${orgId}/system-backup/objects`,
      {},
      token
    ),

  restoreSystem: (orgId: string, key: string, token: string) =>
    apiFetch<void>(
      `/api/v1/orgs/${orgId}/system-backup/restore`,
      { method: "POST", body: JSON.stringify({ key }) },
      token
    ),
}

File: apps/cli/internal/client/backups.go
package client

type BackupConfig struct {
	ID                   string `json:"id"`
	ServiceID            string `json:"service_id"`
	StorageIntegrationID string `json:"storage_integration_id"`
	Schedule             string `json:"schedule"`
	RetentionDays        int    `json:"retention_days"`
	PathPrefix           string `json:"path_prefix"`
	Enabled              bool   `json:"enabled"`
}

type CreateBackupConfigBody struct {
	StorageIntegrationID string `json:"storage_integration_id"`
	Schedule             string `json:"schedule"`
	RetentionDays        int    `json:"retention_days,omitempty"`
	PathPrefix           string `json:"path_prefix,omitempty"`
}

type UpdateBackupConfigBody struct {
	Schedule      *string `json:"schedule,omitempty"`
	RetentionDays *int    `json:"retention_days,omitempty"`
	PathPrefix    *string `json:"path_prefix,omitempty"`
	Enabled       *bool   `json:"enabled,omitempty"`
}

type BackupObject struct {
	Key          string `json:"key"`
	Size         int64  `json:"size"`
	LastModified string `json:"last_modified"`
}

type SystemBackupConfig struct {
	ID                   string `json:"id"`
	StorageIntegrationID string `json:"storage_integration_id"`
	Schedule             string `json:"schedule"`
	RetentionDays        int    `json:"retention_days"`
	PathPrefix           string `json:"path_prefix"`
	Enabled              bool   `json:"enabled"`
}

type UpsertSystemBackupBody struct {
	StorageIntegrationID string `json:"storage_integration_id"`
	Schedule             string `json:"schedule"`
	RetentionDays        int    `json:"retention_days,omitempty"`
	PathPrefix           string `json:"path_prefix,omitempty"`
	Enabled              bool   `json:"enabled"`
}

func backupBase(orgID, projectID, serviceID string) string {
	return "/api/v1/orgs/" + orgID + "/projects/" + projectID + "/services/" + serviceID + "/backups"
}

func (c *Client) ListBackupConfigs(orgID, projectID, serviceID string) ([]BackupConfig, error) {
	resp, err := c.do("GET", backupBase(orgID, projectID, serviceID), nil)
	if err != nil {
		return nil, err
	}
	return decode[[]BackupConfig](resp)
}

func (c *Client) CreateBackupConfig(orgID, projectID, serviceID string, body CreateBackupConfigBody) (*BackupConfig, error) {
	resp, err := c.do("POST", backupBase(orgID, projectID, serviceID), body)
	if err != nil {
		return nil, err
	}
	return decodePtr[BackupConfig](resp)
}

func (c *Client) UpdateBackupConfig(orgID, projectID, serviceID, id string, body UpdateBackupConfigBody) (*BackupConfig, error) {
	resp, err := c.do("PATCH", backupBase(orgID, projectID, serviceID)+"/"+id, body)
	if err != nil {
		return nil, err
	}
	return decodePtr[BackupConfig](resp)
}

func (c *Client) DeleteBackupConfig(orgID, projectID, serviceID, id string) error {
	return c.doNoContent("DELETE", backupBase(orgID, projectID, serviceID)+"/"+id)
}

func (c *Client) TriggerBackup(orgID, projectID, serviceID, id string) error {
	return c.doNoContent("POST", backupBase(orgID, projectID, serviceID)+"/"+id+"/trigger")
}

func (c *Client) ListBackupObjects(orgID, projectID, serviceID, id string) ([]BackupObject, error) {
	resp, err := c.do("GET", backupBase(orgID, projectID, serviceID)+"/"+id+"/objects", nil)
	if err != nil {
		return nil, err
	}
	return decode[[]BackupObject](resp)
}

func (c *Client) RestoreBackup(orgID, projectID, serviceID, id, key string) error {
	resp, err := c.do("POST", backupBase(orgID, projectID, serviceID)+"/"+id+"/restore", map[string]string{"key": key})
	if err != nil {
		return err
	}
	resp.Body.Close()
	return nil
}

func (c *Client) GetSystemBackup(orgID string) (*SystemBackupConfig, error) {
	resp, err := c.do("GET", "/api/v1/orgs/"+orgID+"/system-backup", nil)
	if err != nil {
		return nil, err
	}
	return decodePtr[SystemBackupConfig](resp)
}

func (c *Client) UpsertSystemBackup(orgID string, body UpsertSystemBackupBody) (*SystemBackupConfig, error) {
	resp, err := c.do("PUT", "/api/v1/orgs/"+orgID+"/system-backup", body)
	if err != nil {
		return nil, err
	}
	return decodePtr[SystemBackupConfig](resp)
}

func (c *Client) TriggerSystemBackup(orgID string) error {
	return c.doNoContent("POST", "/api/v1/orgs/"+orgID+"/system-backup/trigger")
}

func (c *Client) ListSystemBackupObjects(orgID string) ([]BackupObject, error) {
	resp, err := c.do("GET", "/api/v1/orgs/"+orgID+"/system-backup/objects", nil)
	if err != nil {
		return nil, err
	}
	return decode[[]BackupObject](resp)
}

func (c *Client) RestoreSystemBackup(orgID, key string) error {
	resp, err := c.do("POST", "/api/v1/orgs/"+orgID+"/system-backup/restore", map[string]string{"key": key})
	if err != nil {
		return err
	}
	resp.Body.Close()
	return nil
}
